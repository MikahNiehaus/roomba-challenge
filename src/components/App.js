import React, { Component } from 'react';
import RoombaInputs from './RoombaInputs';
import CompassAndGridWrapper from './CompassAndGridWrapper';
import './App.css'

class App extends Component {
    constructor() {
        super();
        this.state = {
            grid: {
                width: 0,
                length: 0
            },
            dirt: {
                x: '',
                y: ''
            },
            roomba: {
                x: 0,
                y: 0
            },
            submitButtonClicked: false
        };
    }

    coordinatesCheckIfMaxY = (y) => {
        return y !== this.state.grid.length;
    };

    coordinatesCheckIfMaxX = (x) => {
        return x !== this.state.grid.width;
    };

    coordinatesCheckIfZero = (xOrY) => {
        return xOrY > 0;
    };

    north = () => {
        const roomba = this.state.roomba;

        if(this.coordinatesCheckIfMaxY(roomba.y + 1)) {
            roomba["y"] += 1;
            this.setState({roomba}, this.tileCleaned());
        }
    };

    east = () => {
        const roomba = this.state.roomba;

        if(this.coordinatesCheckIfMaxX(roomba.x + 1)) {
            roomba["x"] += 1;
            this.setState({roomba}, this.tileCleaned());
        }
    };

    south = () => {
        const roomba = this.state.roomba;

        if(this.coordinatesCheckIfZero(roomba.y)) {
            roomba["y"] -= 1;
            this.setState({roomba}, this.tileCleaned());
        }
    };

    west = () => {
        const roomba = this.state.roomba;

        if(this.coordinatesCheckIfZero(roomba.x)) {
            roomba["x"] -= 1;
            this.setState({roomba}, this.tileCleaned());
        }
    };

    handleChange = (event) => {
        const {name, value, placeholder} = event.target;
        const grid = this.state.grid;
        const dirt = this.state.dirt;
        const roomba = this.state.roomba;

        if(placeholder.includes("Grid")) {
            grid[name] = parseInt(value, 10);

            this.setState({
                grid
            });
        }
        if(placeholder.includes("Dirt")) {
            dirt[name] = parseInt(value, 10);

            this.setState({
                dirt
            });
        }
        if(placeholder.includes("Roomba")) {
            roomba[name] = parseInt(value, 10);

            this.setState({
                roomba
            })
        }
    };

    handleSubmit = () => {
        this.tileCleaned();
            this.setState({
                submitButtonClicked: true
            });
    };

    tileCleaned() {
        if(JSON.stringify(this.state.roomba) === JSON.stringify(this.state.dirt)) {
            this.setState({
                dirt: "game over"
            })
        }
    }

    render() {
        return (
            <div>
                <div className="header">
                    <span>Roomba</span>
                </div>
                {this.state.submitButtonClicked ?
                    <CompassAndGridWrapper
                        appState={this.state}
                        north={this.north}
                        east={this.east}
                        south={this.south}
                        west={this.west}
                    /> :
                    <center>
                    <RoombaInputs
                        // onChange={this.handleChange}
                        // onSubmit={this.handleSubmit}
                        // grid={this.state.grid}
                    /></center>
                }
            </div>
        );
    }
}

export default App;