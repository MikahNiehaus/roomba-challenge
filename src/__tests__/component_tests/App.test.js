import React from 'react';
import {shallow} from 'enzyme';
import App from '../../components/App';

let app;

const state = {
    grid: {
        length: 0,
        width: 0
    },
    dirt: {
        x: '',
        y: ''
    },
    roomba: {
        x: 0,
        y: 0
    },
    submitButtonClicked: false
};

const roomDimensionsEvent = {
    target: {
        name: 'length',
        value: '3',
        placeholder: 'room'
    }
};

const dirtCoordinatesEvent = {
    target: {
        name: 'y',
        value: '10',
        placeholder: 'place your dirt co-ordinates here'
    }
};

describe('App', () => {

    app = shallow(<App/>);

    beforeEach(() => {
        app = shallow(<App/>);
    });

    it('renders correctly', () => {
        expect(app).toMatchSnapshot();
    });

    describe('Initializes state', () => {
        it('initializes state correctly', () => {
            expect(app.state()).toEqual(state);
        });
    });

    describe('Coordinate checkers', () => {

        describe('coordinatesCheckIfZero returns a boolean depending on coordinates set in state', () => {
            it('returns true if argument as object is greater than 0', () => {
                expect(app.instance().coordinatesCheckIfZero(1)).toEqual(true);
            });

            it('returns false if argument as object is less than or equal to 0', () => {
                expect(app.instance().coordinatesCheckIfZero(0)).toEqual(false);
            });
        });

        describe('coordinatesCheckIfMaxX returns a boolean depending on coordinates set in state', () => {
            it('returns false if argument equals the current grid length', () => {
                expect(app.instance().coordinatesCheckIfMaxX(0)).toEqual(false);
            });
        });

        describe('coordinatesCheckIfMaxY returns a boolean depending on coordinates set in state', () => {
            it('returns true if argument does not equal the current grid length', () => {
                expect(app.instance().coordinatesCheckIfMaxY(20)).toEqual(true);
            });
        });
    });

    describe('Updating state successfully', () => {

        describe('handleChange', () => {
            it('updates the state correctly with a room dimensions event', () => {
                app.instance().handleChange(roomDimensionsEvent);
                expect(app.state().grid.length).toEqual(3);
            });

            it('updates the state correctly with a dirt co-ordinates event', () => {
                app.instance().handleChange(dirtCoordinatesEvent);
                expect(app.state().dirt.y).toEqual(10);
            });
        });

        describe('handleSubmit', () => {
            it('updates the state correctly when submit button is clicked', () => {
                app.instance().handleSubmit();
                expect(app.state().submitButtonClicked).toEqual(true);
            })
        });

        describe('compass directions for Roomba', () => {

            describe('north', () => {
                it('increases the y value on Roomba by 1 value', () => {
                    app.instance().north();
                    expect(app.state().roomba.y).toEqual(1)
                });
            });

            describe('east', () => {
                it('increases the x value on Roomba by 1 value', () => {
                    app.instance().east();
                    expect(app.state().roomba.x).toEqual(1)
                });
            });

            describe('west', () => {
                it('decreases the x value on Roomba by 1 value', () => {
                    app.instance().east();
                    app.instance().west();
                    expect(app.state().roomba.x).toEqual(0)
                });
            });

            describe('south', () => {
                it('decreases the y value on Roomba by 1 value', () => {
                    app.instance().north();
                    app.instance().south();
                    expect(app.state().roomba.y).toEqual(0)
                });
            });
        });

        describe('cleaning a tile', () => {
            app.state().dirt = {x: 0, y: 0};
            app.instance().tileCleaned();
            expect(app.state().dirt).toEqual("game over");
        });
    });

    describe('app renders depending on whether the submit button is clicked', () => {
        describe('when the button has not been clicked', () => {
            expect(app.find('RoombaInputs').exists()).toBe(true);
            expect(app.find('CompassAndGridWrapper').exists()).toBe(false);
        });
    });

});