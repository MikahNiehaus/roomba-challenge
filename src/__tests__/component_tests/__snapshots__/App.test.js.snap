// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Roomba
</h1>,
        <RoombaInputs
          grid={
                    Object {
                              "length": 0,
                              "width": 0,
                            }
          }
          onChange={[Function]}
          onSubmit={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Roomba",
        },
        "ref": null,
        "rendered": "Roomba",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "grid": Object {
            "length": 0,
            "width": 0,
          },
          "onChange": [Function],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Roomba
</h1>,
          <RoombaInputs
            grid={
                        Object {
                                    "length": 0,
                                    "width": 0,
                                  }
            }
            onChange={[Function]}
            onSubmit={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Roomba",
          },
          "ref": null,
          "rendered": "Roomba",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "grid": Object {
              "length": 0,
              "width": 0,
            },
            "onChange": [Function],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
